Refactor FormListener

redi-wadash opened the pull request:
Basically moved all member variables into member function that returns the variable.

redi-wadash committed:
Refactor FormListener

redi-wadash commented on client/src/main/scala/com/redi/scalademo/presentation/FormListener.scala:
```diff
@@ -95,34 +75,66 @@ class FormListener(
+    var shouldValidate = false
+    for (formControlData: js.Dictionary[String] ← serializedFormData) {
+      if (formControlData(Name) == Validate) {
+        shouldValidate &= formControlData(Value).nonEmpty
```
@redi-russela it used to be ```=``` but I guessed you meant ```&=```. Correct me if I got it wrong.

redi-russela replied:
I'm on the fence whether we should use `&=` over `=` since it will never actually do anything different. There's only one element that will hit this code path.

redi-wadash replied:
I see. I changed how it checks if the form should be validated on client side in new code. Please review it and see if you like it better.

redi-russela commented on client/src/main/scala/com/redi/scalademo/presentation/FormListener.scala:
```diff
@@ -95,34 +75,66 @@ class FormListener(
+    }
+  }
+
+  private def augend: String = getValueOrBlank(Augend)
```
Probably should wrap the body in `{}` since it's more similar to Java syntax.

redi-wadash replied:
Good point. I wrapped it with ```{}```.

redi-russela commented on client/src/main/scala/com/redi/scalademo/presentation/FormListener.scala:
```diff
@@ -95,34 +75,66 @@ class FormListener(
+    var shouldValidate = false
+    for (formControlData: js.Dictionary[String] ← serializedFormData) {
+      if (formControlData(Name) == Validate) {
+        shouldValidate &= formControlData(Value).nonEmpty
```
I'm on the fence whether we should use `&=` over `=` since it will never actually do anything different.

redi-russela commented on client/src/main/scala/com/redi/scalademo/presentation/FormListener.scala:
```diff
@@ -64,27 +59,12 @@ class FormListener(
-  }
-
-  private def processValidationFormData(): Unit = {
+  private def resetStyles(): Unit = {
```
This doesn't reset all styles, though. It's only validation failures. Result box has its own styling concerns.

redi-wadash replied:
Thanks. I renamed it to ```resetValidationErrors()```.

redi-russela commented on client/src/main/scala/com/redi/scalademo/presentation/FormListener.scala:
```diff
@@ -95,34 +75,66 @@ class FormListener(
+  }
+
+  private def calculableFormData: js.Array[js.Dictionary[String]] = {
+    var calculableFormData = js.Array[js.Dictionary[String]]()
```
Looks like it could be `val`.

redi-russela commented on client/src/main/scala/com/redi/scalademo/presentation/FormListener.scala:
```diff
@@ -95,34 +75,66 @@ class FormListener(
-              .addClass(InfoClass)
-        }
+    val futureResult: Future[ValidatedResult[BigDecimal]] =
+      client[Calculator].add(augend, addend).call()
```
Trying to put myself in the mindset of NY folks, here. Before, getting `augend` and `addend` required only one iteration through `calculableFormData` in `validateAndSetStyles()` (formerly `validateAndExtractNumbers()`). I think I'd like to retain the efficiency, small as though it may be. Perhaps return a tuple or a special-purpose nested class instance?

redi-wadash replied:
I changed it so that it stores the elements to variables. It is more efficient and IMO more readable.

redi-wadash committed:
Store elements to variables

redi-wadash committed:
Refactor

redi-wadash committed:
Refactor

redi-wadash committed:
Merge branch 'master'

Conflicts:
	client/src/main/scala/com/redi/scalademo/presentation/FormListener.scala

redi-wadash committed:
Print on validation

This is to demo where the validation is happening in addition to network log.

redi-wadash commented on shared/src/main/scala/com/redi/scalademo/business/NumericStringValidator.scala:
```diff
@@ -5,6 +5,7 @@ import scala.util.Try
 class NumericStringValidator {
 
   def isValidNumber(string: String): Boolean = {
+    println(s"Validating if '$string' is a valid number")
```
I thought, in addition to network log, printing when the validation is happening is a good way to demo where the validation is actually happening.

redi-wadash commented on client/src/main/scala/com/redi/scalademo/presentation/FormListener.scala:
```diff
@@ -55,80 +49,96 @@ class FormListener(
-  private val Summand = "summand"
-
-  private val Name = "name"
-  private val Value = "value"
```
To demo smoothly and minimize going back-and-forth in the code, I removed constants that are only used once.

redi-wadash replied:
Actually, I realized these constants are used by server too. I'll move them to shared project.

redi-wadash committed:
Extract form control names to shared project

redi-tjandr commented on client/src/main/scala/com/redi/scalademo/presentation/FormListener.scala:
```diff
@@ -1,11 +1,12 @@
 package com.redi.scalademo.presentation
```
I feel like this class would be more appropriately named "CalculatorFormListener"

redi-wadash replied:
Renamed

redi-wadash committed:
Rename FormListener to CalculatorFormListener

redi-wadash commented on server/src/main/scala/com/redi/scalademo/business/DefaultCalculator.scala:
```diff
@@ -9,10 +9,10 @@ class DefaultCalculator(
     }
     if (!validator.isValidNumber(addend)) {
-      failures += ValidationFailure("addend", "addend must be a number")
+      failures += ValidationFailure(FormControlNames.Addend, "addend must be a number")
```
One of the advantages of sharing code; we can share constants.

redi-wadash commented on client/src/main/scala/com/redi/scalademo/application/ClientModule.scala:
```diff
@@ -1,10 +1,10 @@
 object ClientModule extends SharedModule {
   lazy val client = wire[Client]
-  lazy val formListener = wire[FormListener]
+  lazy val calculationFormListener = (calculationFormId: String) => wire[CalculatorFormListener]
```
A way to declare factory using macwire.

It can be ```def calculationFormListener(calculationFormId: String) = wire[CalculatorFormListener]``` too. That's preference and I don't mind either.

redi-russela committed:
Change symbolized operators to ASCII versions.

This is simply for consistency's sake, and it's safer to use the version that
can be typed on a keyboard without an IDE's assistance unless/until we decide
we like the symbols better.

redi-russela committed:
Minor spacing changes in CalculatorFormListener.

redi-russela commented:
@redi-wadash Made some minor changes to reduce back-and-forth. This looks good to me, if my changes are fine to you!
